File: programmers/level_1/42889.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool check[501];

vector<int> solution(int N, vector<int> stages) {
    vector<int> answer;

    vector<pair<double, int>> v;
    sort(stages.begin(), stages.end());
    for(int i = 0; i < stages.size();) {
        int curStage = stages[i];
        if(curStage == N+1) {
            break;
        }
        int nxtStage = curStage;
        int count = 0;
        do {
            count++;
            nxtStage = stages[i+count];
        } while(curStage == nxtStage);
        double fail = (double)count/(double)(stages.size() - i);
        v.push_back({-fail, curStage});
        i = i + count;
    }

    sort(v.begin(), v.end());

    for(auto i: v) {
        answer.push_back(i.second);
    }

    return answer;
}

File: kotlin_ps/src/main/kotlin/programmers_42889/42889.kt
package programmers_42889

class Solution {
    fun solution(N: Int, stages: IntArray): IntArray {
        var answer = intArrayOf()

        val notCleared = MutableList<Int>(N) { _ -> 0 }
        val approach = MutableList<Int>(N) { _ -> 0 }
        for (i in notCleared.indices) {
            notCleared[i] = stages.count {
                it == i + 1
            }
        }
        for (i in approach.indices) {
            approach[i] = stages.count {
                it >= i + 1
            }
        }

        val failure = mutableListOf<Pair<Double, Int>>()
        for (i in notCleared.indices) {
            failure.add(Pair(notCleared[i].toDouble() / approach[i].toDouble(), i + 1))
        }
        failure.sortWith(compareBy<Pair<Double, Int>> { -it.first }.thenBy { it.second })
        answer = failure.map { it.second }.toIntArray()

        return answer
    }
}